proof/access-control/RISCV64/ArchAccess_AC.thy

```diff
@@ -71,19 +78,73 @@ lemma auth_ipc_buffers_tro[Access_AC_assms]:
                split: cap.split_asm arch_cap.split_asm if_split_asm bool.splits)
 
 lemma trasids_trans[Access_AC_assms]:
-  "\<lbrakk> (\<forall>x a. integrity_asids aag subjects x a s s');
-     (\<forall>x a. integrity_asids aag subjects x a  s' s'') \<rbrakk>
-     \<Longrightarrow> (\<forall>x a. integrity_asids aag subjects x a s s'')"
-  by clarsimp metis
+  "\<lbrakk> \<forall>x a. integrity_asids_2 aag subjects x a as as' ao ao';
+     \<forall>x a. integrity_asids_2 aag subjects x a as' as'' ao' ao'' \<rbrakk>
+     \<Longrightarrow> \<forall>x a. integrity_asids_2 aag subjects x a as as'' ao ao''"
+  by (clarsimp simp: integrity_asids_def) metis
 
 lemma integrity_asids_refl[Access_AC_assms, simp]:
   "integrity_asids aag subjects x a s s"
-  by simp
+  by (simp add: integrity_asids_def)
 
 lemma integrity_asids_update_autarch[Access_AC_assms]:
-  "\<lbrakk> \<forall>x a. integrity_asids aag {pasSubject aag} x a st s; is_subject aag ptr \<rbrakk>
-     \<Longrightarrow> \<forall>x a. integrity_asids aag {pasSubject aag} x a st (s\<lparr>kheap := (kheap s)(ptr \<mapsto> obj)\<rparr>)"
-  by (auto simp: opt_map_def)
+  "\<lbrakk> integrity_asids_2 aag subjects x a as as' ao ao'; pasObjectAbs aag ptr \<in> subjects \<rbrakk>
+     \<Longrightarrow> integrity_asids_2 aag subjects x a as as' ao (ao'(ptr := ako))"
+  by (auto simp: integrity_asids_def opt_map_def)
+
+lemma integrity_hyp_triv[intro!,simp]:
+  "integrity_hyp_2 aag subjects x ms ms' as as' ao ao'"
+  by (simp add: integrity_hyp_def)
+
+lemma integrity_fpu_triv[intro!,simp]:
+  "integrity_fpu_2 aag subjects x as as' kh kh'"
+  by (simp add: integrity_fpu_def)
+
+lemma integrity_hyp_wp_triv:
+  "\<lbrace>P\<rbrace> c \<lbrace>\<lambda>_ s. integrity_hyp aag subjects x st (f s)\<rbrace>"
+  by wpsimp
+
+lemma integrity_fpu_wp_triv:
+  "\<lbrace>P\<rbrace> c \<lbrace>\<lambda>_ s. integrity_fpu aag subjects x st (f s)\<rbrace>"
+  by wpsimp
+
+lemma assumption2:
+  "P \<Longrightarrow> Q \<Longrightarrow> P"
+  by simp
+
+lemma integrity_hyp_eq_triv:
+  "integrity_hyp_2 a b c d e f g h i = integrity_hyp_2 r s t u v w x y z"
+  by simp
+
+lemma integrity_fpu_eq_triv:
+  "integrity_fpu_2 a b c d e f g = integrity_fpu_2 t u v w x y z"
+  by simp
+
+lemmas integrity_arch_triv =
+  integrity_hyp_wp_triv
+  integrity_fpu_wp_triv
+  integrity_hyp_triv
+  assumption2[OF integrity_hyp_triv]
+  assumption2[OF assumption2[OF integrity_hyp_triv]]
+  assumption2[OF assumption2[OF assumption2[OF integrity_hyp_triv]]]
```

Comment:
Can you explain what is going on with these instances? What are we trying to achieve?

I have the feeling that we can get away without assumption2, but I'm not actually clear on what final lemmas we are trying to build.

Reply:
Couldn't get `fact` to solve unless the assumption count matched. Could use `rule` or such instead

Reply:
Ah, interesting. Too used to  `assumption`. I think it would make sense to use the `assumption` method instead of `fact` in this case. We do want the interface to match up, but just having fewer assumptions shouldn't receive that much punishment.

Reply:
The `assumption` method wouldn't actually work. Using OF means the rules end up looking like:
`P ==> integrity_hyp`
`P ==> Q ==> integrity_hyp`
etc.

Reply:
I do agree that this is brittle and feels redundant. If we let ourselves stray away from `fact`, we could have `solves (rule integrity_hyp_triv)`